plugins/gradle/tooling-extension-impl/src/org/jetbrains/plugins/gradle/tooling/tasks/DependenciesReport.groovy

```diff
@@ -54,191 +53,148 @@ class DependenciesReport extends DefaultTask {
+          node = new UnknownDependencyNode(id, componentId.displayName)
+        }
+        node.resolutionState = RESOLVED
+        if (!dependency.selected.selectionReason.descriptions.isEmpty()) {
+          node.selectionReason = dependency.selected.selectionReason.descriptions.last().description
+        }
+        added.put(id, node)
+        def iterator = dependency.selected.dependencies.iterator()
+        while (iterator.hasNext()) {
+          DependencyResult child = iterator.next()
+          node.dependencies.add(toNode(child, added, idGenerator, projectNameFunction))
         }
+      } else if (dependency instanceof UnresolvedDependencyResult) {
+        node = new UnknownDependencyNode(id, dependency.attempted.displayName)
+        node.resolutionState = UNRESOLVED
+        added.put(id, node)
+      } else {
+        throw new IllegalStateException("Unsupported dependency result type: $dependency")
       }
 
-      return out
+      return node
     }
-  }
-
-  private static class CustomRenderableDependencyResult extends AbstractRenderableDependencyResult {
-    private final ResolvedDependencyResult dependency
 
-    CustomRenderableDependencyResult(ResolvedDependencyResult dependency) {
-      this.dependency = dependency
-    }
+    private static class IdGenerator {
+      private Map<Object, Long> idMap = new HashMap<>()
+      private AtomicLong nextId
 
-    ResolutionState getResolutionState() {
-      return this.dependency.isConstraint() ? ResolutionState.RESOLVED_CONSTRAINT : ResolutionState.RESOLVED
-    }
+      IdGenerator(AtomicLong nextId) {
+        this.nextId = nextId
+      }
 
-    ComponentIdentifier getActual() {
-      return this.dependency.getSelected().getId()
-    }
+      long getId(Object key) {
+        def existingId = idMap.get(key)
+        if (existingId != null) {
+          return existingId
+        }
 
-    ComponentSelector getRequested() {
-      return this.dependency.getRequested()
+        def newId = nextId.incrementAndGet()
+        idMap.put(key, newId)
+        return newId
+      }
     }
+  }
 
-    ComponentSelectionReason getSelectionReason() {
-      return dependency.selected.selectionReason
-    }
+  static class ProjectNameFunction {
+    def is45OrNewer = GradleVersion.current() >= GradleVersion.version("4.5")
 
-    Set<RenderableDependency> getChildren() {
-      Set<RenderableDependency> out = new LinkedHashSet()
-      for (def dependencyResult in dependency.getSelected().dependencies) {
-        if (dependencyResult instanceof UnresolvedDependencyResult) {
-          out.add(new RenderableUnresolvedDependencyResult((UnresolvedDependencyResult)dependencyResult))
-        }
-        else {
-          out.add(new CustomRenderableDependencyResult((ResolvedDependencyResult)dependencyResult))
-        }
-      }
-      return out
+    String fun(ProjectComponentIdentifier identifier) {
+      return is45OrNewer ? identifier.projectName : identifier.projectPath
```

Comment:
Why don't you use `project.displayName` which has always existed and returns the full path, even in composite builds?